src/Simulation/Simulators/SparseSimulator/Native/SparseSimulator.h

```diff
@@ -0,0 +1,846 @@
+							result ^= _H_queue[qubits[i]];
+						}
+						isAllZ = false;
+						isEmpty = false;
+						break; 
+					case Gates::Basis::PauliX:
+						// HX = ZH
+						if (_H_queue[qubits[i]]){
+							axes[i] = Gates::Basis::PauliZ;
+						} else {
+							isAllZ = false;
+						}
+						isEmpty = false;
+						break;
+					case Gates::Basis::PauliZ:
+						// HZ = XH
+						if (_H_queue[qubits[i]]){
+							axes[i] = Gates::Basis::PauliX;
+							isAllZ = false;
+						}
+						isEmpty = false;
+						break;
+					default:
+						break;
+				}
+			}
+			if (isEmpty) {
+				return;
+			}
+			// Z assertions are like phase gates
+			// If it's in release mode, it will queue them
+			// as a phase/permutation gate
+			// This means if an assert fails, it will fail
+			// at some future point, not at the point of failure
+			#if NDEBUG
+				if (isAllZ) {
+					_queued_operations.push_back(operation(OP::Assert, qubits, result));
+					return;
+				}
+			#endif
+			// X or Y assertions require execution
+			_execute_queued_ops(qubits, OP::PermuteLarge);
+			_quantum_state->Assert(axes, qubits, result);
+		}
+
+		// Returns the probability of a given measurement in a Pauli basis
+		// by decomposing each pair of computational basis states into eigenvectors
+		// and adding the coefficients of the respective components
+		double MeasurementProbability(std::vector<Gates::Basis> const& axes, std::vector<logical_qubit_id> const& qubits) {
+			_execute_queued_ops(qubits, OP::Ry);
+			return _quantum_state->MeasurementProbability(axes, qubits);
+		}
+
+
+
+		bool Measure(std::vector<Gates::Basis> const& axes, std::vector<logical_qubit_id> const& qubits){
+			_execute_queued_ops(qubits, OP::Ry);
+			bool result = _quantum_state->Measure(axes, qubits);
+			// Switch basis to save space
+			// Idea being that, e.g., HH = I, but if we know 
+			// that the qubit is in the X-basis, we can apply H
+			// and execute, and this will send that qubit to all ones
+			// or all zeros; then we leave the second H in the queue
+			// Ideally we would also do that with Y, but HS would force execution,
+			// rendering it pointless
+			std::vector<logical_qubit_id> measurements;
+			for (int i =0; i < axes.size(); i++){
+				if (axes[i]==Gates::Basis::PauliX){
+					H(qubits[i]);
+					measurements.push_back(qubits[i]);
+				} 
+			}
+			_execute_queued_ops(measurements, OP::H);
+			// These operations undo the previous operations, but they will be 
+			// queued 
+			for (int i =0; i < axes.size(); i++){
+				if (axes[i]==Gates::Basis::PauliX){
+					H(qubits[i]);
+				} 
+			}
+			return result;
+		}
+
+		// Returns the amplitude of a given bitstring
+		amplitude probe(std::string label) {
+			_execute_queued_ops();
+			return _quantum_state->probe(label);
+		}
+
+		std::string Sample() {
+			_execute_queued_ops();
+			return _quantum_state->Sample();
+		}
+
+		// Dumps the state of a subspace of particular qubits, if they are not entangled
```

Comment:
It passes back "false" to C#, which outputs the error message [here](https://github.com/sam-jaques/qsharp-runtime/blob/5470276e43b31dae7f5d6e20d03a2f8a8695942e/src/Simulation/Simulators/SparseSimulator/SparseSimulatorCS/Probes.cs#L122) to the output channel where it would otherwise print an unentangled state.